buildDotnetModule: use setup hooks

IvarWithoutBones opened the pull request:
###### Motivation for this change

As suggested by @jtojnar in #153364. See the commit descriptions.

This is heavily inspired by `pkgs/build-support/rust`, so thanks to everyone who worked on that!

cc @prusnak.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

IvarWithoutBones committed:
buildDotnetModule: use setup hooks

This is a much more flexible way of doing things, as we can adopt and
reuse these hooks for various tasks. Syntax highlighting now also works
way better for me, which is a nice bonus :)

IvarWithoutBones commented on pkgs/build-support/build-dotnet-module/hooks/dotnet-configure-hook.sh:
```diff
@@ -0,0 +1,32 @@
+        parallelFlag="--disable-parallel"
+    fi
+
+    export HOME=$(mktemp -d)
```
I'm not quite sure what the best place to set this is, all hooks depend on it.

SuperSandro2000 replied:
I would just export it at the earliest possible point we need it. We don't need to create a giant mess of hooks.

aanderse commented:
Sorry @IvarWithoutBones I don't think I have anything to offer here :disappointed: I wonder if @jonringer is able/willing to review?

jonringer commented:
you rang?

jonringer commented:
I'm fine with using this, most other ecosystems (at least python, rust) do something similar.

jonringer reviewed (approved):
gave it a quick glance, and looks like idiomatic setuphook usage

```
https://github.com/NixOS/nixpkgs/pull/155257

15 packages built:
ArchiSteamFarm alttpr-opentracker baget btcpayserver depotdownloader discordchatexporter-cli jackett nbxplorer netcoredbg opentabletdriver osu-lazer pinta python-language-server ryujinx wasabibackend
```

SuperSandro2000 commented on pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh:
```diff
@@ -0,0 +1,37 @@
+                -p:BuildInParallel=$parallelBuildFlag \
+                -p:ContinuousIntegrationBuild=true \
+                -p:Deterministic=true \
+                -p:Version="${version}" \
```
What happens if just name is set and version is left empty?

SuperSandro2000 commented on pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh:
```diff
@@ -0,0 +1,37 @@
+    runHook preBuild
+
+    if [ ! -z "${enableParallelBuilding-}" ]; then
+        maxCpuFlag="$NIX_BUILD_CORES"
+        parallelBuildFlag="true"
```
Do we local those variables in other hooks?

IvarWithoutBones replied:
`$maxCpuFlag` is used in `dotnet-check-hook`, and the value stays the same. The second one only seems to be used while actually building.

SuperSandro2000 commented on pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh:
```diff
@@ -0,0 +1,42 @@
+
+    if [ ! -z "${executables-}" ]; then
+        for executable in $executables; do
+            execPath="$out/lib/${pname}/$executable"
```
I think there should be a way to override this if pname does not match the lib name.

IvarWithoutBones replied:
Good idea, that could be useful in scenarios where `$out/lib/$pname` is already used as well. I've added the `installPath` flag :+1:

lilyinstarlight replied:
So, um, `installPath` was added as a flag to `buildDotnetModule`, but as I learned today, it doesn't seem to be used anywhere or do anything 😅

I imagine it probably got lost during refactoring, but I've opened #251267 about it

IvarWithoutBones commented:
Resolved merge conflicts :+1:

IvarWithoutBones committed:
buildDotnetModule: enable RestoreUseStaticGraphEvaluation

This should speed up restore times a fair bit, especially for bigger
projects. Roslyn also has it enabled by default already, so I don't
expect any breakages from it.

IvarWithoutBones commented:
Is there anything blocking this? I'd love to get it merged, i have some changes (almost) ready that depend on it

SuperSandro2000 commented:
I don't really have the time to look into such a big diff right now.

winterqt reviewed (approved):
Diff LGTM.

IvarWithoutBones commented:
Thanks!
